Dedicated shared message contracts documentation (#5622)
Particular/docs.particular.net: nservicebus/messaging/evolving-contracts.md

This approach requires all senders/publishers to be updated first. Due to message inheritance, receivers continue to process the original message version.

## Removing contracts or properties

1. Update all receivers/publishers to no longer use the property or message type to be removed
1. Remove the contract or property from the contracts
1. Update senders/publishers to the new contracts assembly.

The Obsolete attribute can be used to mark properties/types to be removed beforehand to give consumers of the contract time to update their code.

## Versioning

* Keep the assembly version on 1.0.0 to avoid assembly loading conflicts when endpoints with different contract versions send messages to each other
* Use file version and/or NuGet package versions to indicate the sematic version of the contracts assembly

## Modifying serialization formats

Another approach for handling breaking changes is to modify serialization formats. Step-by-step guidance is provided in the [transition serialization formats](/samples/serializers/transitioning-formats/) sample.

--------------------------------------------------------------------
Different proposal:

## Techniques

> There are different techniques to evolve message contracts, each bringing it's own advantages and disadvantages. When selecting a technique it's important to carefully plan the migration strategy, the needs of message receivers/subscribers and message senders/publishers.

Reviewer: ```suggestion
There are different techniques to evolve message contracts, each one with advantages and disadvantages. When selecting a technique it's important to carefully plan the migration strategy, the needs of receivers/subscribers and senders/publishers.
```

### New message type

Ship multiple versions of a message contract in the same assembly by creating a new message type for the new contract version. The new type typically indicates its relationship via its name, e.g.

```
// first version of the message contract

> public class CreateOrder : ICommand {
>     ...
> }

Reviewer: ```suggestion
public class CreateOrder : ICommand
{
    ...
}
```

Shall we use the typical C# formatting style?

// new version of the message contract
public class CreateOrderV2 : Icommand{
    ...
}
```

> * Consumers can be updated to the new contract version independent from updating the contract assembly, giving more flexibility in planning the upgrade process.

Reviewer: ```suggestion
* Consumers can be updated to the new version independently from updating the contract assembly, giving more flexibility in planning the upgrade process.
```

> * Allows more flexibility about the contract changes as no type-level compatibility has to be ensured.

Reviewer: ```suggestion
* Allows more flexibility about the contract changes as no type-level compatibility must be ensured.
```

> When updating commands/messages:
> * Ensure the receiver of the message has a handler for both types to ensure it can process messages that are still in the queue or coming from endpoints still using the old contract.

Reviewer: ```suggestion
When updating commands/messages, make sure receivers have a handler for both types. This ensures they can process in-flight messages or messages sent by endpoints still using the old contract.
```

> When updating events:
> * Update all subscribers to have handlers (and subscribe) to both events before changing the publisher.

Reviewer: ```suggestion
When updating events, update all subscribers to have handlers (and subscribe) to both events before changing publishers.
```

### Use inheritance

Ship multiple versions of a message contract in the same assembly by using inheritance. The new contract version inherits from the previous version, gaining all its parent's properties.

```
// first version of the message contract
public class OrderCreatedEvent : IEvent {
    ...
}

public class OrderCreatedEventV2 : OrderCreatedEvent {
    ...
}
```

* Can only be used to make additive changes
* Flattening the inheritance chain / removing older versions is more difficult compared to the other options

> * Something about multiple/all message handlers being invoked

Reviewer: Something 😄

> * Something about routing complexity?

Reviewer: Something 😄

> When updating commands/messages:
> * Update all senders to the new version before adding a new message handler for the new version. This might delay the time till the new message contract can be implemented by the receiver significantly.

Reviewer: ```suggestion
When updating commands/messages, update all senders to the new version before adding new message handlers for the new version to receivers. This might delay the time till the new message contract can be implemented by receivers significantly.
```

> When updating events:
> * Update the publisher first to publish only the latest version of the event. Subscribers using the old message contract will also receive the event but process it the same as the previous version of the contract.

Reviewer: ```suggestion
When updating events, update publishers first to publish only the latest version of the event. Subscribers using the old message contract will also receive the event but process it as if it was the previous version.
```

> ### Modify contract (adding data)

Reviewer: ```suggestion
### Adding data to existing contracts
```

When modifying an existing contract, contract consumers updating to the latest contract version will only know this version of the contract. When only making additive changes, the serialization behavior can be used to ensure backwards compatibility with older versions of the contract.

```
// first version of the message contract
public class CreateOrder : ICommand {
    ...
    // this property has been added in the latest version:
    public int? SomeProperty { get; set; }
}
```

* Endpoints using the old version of the contract can process newer versions since the additional data is just being ignored

> * The new data should be using nullable types to ensure endpoints can distinguish between missing data and intentionally set default values (e.g., `0` when using `int`).

Reviewer: I have a feeling this is a mess to handle in reality. It's like when you have checks in the code to determine the meaning of a SQL field based on its nullability.

* Endpoints using the new version of the contract need to expect the new data to be missing if there are still messages in the queue when updating, or if endpoints can send/publish using the old contract.

> When updating commands/messages:
> * The receiver of the message should be able to handle messages that have the new property missing. If that is not possible, all senders must be updated first, potentially delaying the time till the new data can be fully used by the receiver.

Reviewer: ```suggestion
When updating commands/messages, the receiver of the message should be able to handle messages that have the new property missing. If that is not possible, all senders must be updated first, potentially delaying the time till the new data can be fully used by the receiver.
```

> When updating events:
> * Update the publisher first. Subscribers using the old contract will simply ignore the additional data and can be updated one at a time.

Reviewer: ```suggestion
When updating events, update publishers first. Subscribers using the old contract will simply ignore the additional data and can be updated one at a time.
```

> ### Modify contract (removing data)

Reviewer: ```suggestion
### Removing data from existing contracts
```

When modifying an existing contract, contract consumers updating to the latest contract version will only know this version of the contract. When removing properties from a contract, extra care has to be taken since this can impact endpoints still referencing the old version of the contract that aren't able to handle the missing data.

When updating commands/messages:
* Update receivers first to ensure they are no longer depending on the property being removed. During this time, the senders might still send the additional data till they are updated.

> When updating events:
> * Update all subscribers first to ensure they are no longer depending on the property being removed. During this time, the publisher still need to provide the data being removed.

Reviewer: ```suggestion
When updating events, update subscribers first to ensure they are no longer depending on the property being removed. During this time, publishers still need to provide the data being removed.
```

* Endpoints updated to the latest message contract might send messages using the new contract to endpoints that are still deployed with the old contract version.

Generally the problem can't be resolved at the infrastructure level; therefore NServiceBus users must analyze their individual systems, consider how they are expected to evolve, and define the strategy which will make the most sense in their particular circumstances.

This article presents basic guidelines for choosing contracts evolution strategy, avoiding common mistakes, and ensuring that contracts will be easy to evolve over time.

TODO: link to new document
Note: Ensure that message contracts follow the general [messages design guidelines](/nservicebus/messaging/messages-events-commands.md#designing-messages).

## Adding new message contracts

When adding new message contracts to a contracts assembly, ensure to update the endpoints receveiving or subscribing to the new contract type first. When updating the sender/publisher first, the messages fail to be processed by endpoints that are not using the new contract assembly.

## Adding data to existing contracts

> Adding additional data to existing contracts is the most common change of contracts. There are differnt approaches available.

Reviewer 2: "Adding data" means adding properties to the contract?

> Adding additional data to existing contracts is the most common change of contracts. There are differnt approaches available.

Author: yes, at least that's how I understand it. The original document used the "adding data" wording.

> ### Create a new contract type

Author: depends on what you'd consider "versioning". It's not assembly/package versioning (that can be used to indicate the change in the contract) but it would be something like this:

```
class OrderCreated : IMessage {
  string MyProperty1 {get;set;}
}

class OrderCreatedV2 : IEvent {
  string MyProperty1 {get;set;}
  string MyProperty2 {get;set;}
}
```

> ### Create a new contract type

Author: (duck typing)

1. Add message handlers for the new contract type to the receiving endpoints. Keep the existing mesage handlers for the "old" contract.
1. Update the sender/publisher to send messages using the new contract type.
1. Once all endpoints have been updated, and no more messages using the old contract type is left in the queue, the old type and associated handlers can be safely removed.

This approaches requires all receivers to be updated first, as they will be aware of the new contract type before upgrading which results in message processing failures.

TODO: what about version namespaces?

### Add new properties to existing contract types

1. Add the new properties to the existing type.
1. Update senders to the new contract version, setting the additional data on the message
1. Update receivers to handle the new contract version.

This approach requires less modifications to the contracts and even allows to update senders and receivers in arbitrary order as long as the receivers can handle the absence of a value for the new property.
TODO: can serialzers handle the absence?  

> It is recommended to use nullable types for the new properties to allow receivers to identify whether it is dealing with an old version of the contract.

Author: I primarily took that statement from the original document without thinking too much about it tbh. Depending on your migration approach (receiver- or sender-first), it might be very useful or not necessary at all. Primitive types are tricky without nullables, e.g. I think we even had (or still have?) some limitations in core that saga correlation properties can't be `0` because in that case we don't know whether the sender didn't set it or whether the sender really wanted it to be 0. Nullables can solve this properly.

### Use inheritance to create a new sub-type

1. Create a new contract type, inheriting from the contract type that should be extended
1. Add the new properties to the new sub-type
1. Update the senders to publish/send the new sub-type
1. Update the receivers to the new sub-type

This approach requires all senders/publishers to be updated first. Due to message inheritance, receivers continue to process the original message version.

## Removing contracts or properties

1. Update all receivers/publishers to no longer use the property or message type to be removed
1. Remove the contract or property from the contracts
1. Update senders/publishers to the new contracts assembly.

The Obsolete attribute can be used to mark properties/types to be removed beforehand to give consumers of the contract time to update their code.

## Versioning

* Keep the assembly version on 1.0.0 to avoid assembly loading conflicts when endpoints with different contract versions send messages to each other

> * Use file version and/or NuGet package versions to indicate the sematic version of the contracts assembly

Author: ![](https://i.kym-cdn.com/entries/icons/facebook/000/006/759/both.jpg)

* Endpoints updated to the latest message contract might still receive messages using the old contract. Senders might still use the old contract, or not all in-flight messages (messages waiting to be consumed in input queues) have been processed before the upgrade.

* Endpoints updated to the latest message contract might send messages, using the new contract, to endpoints still based on the old contract version.

Generally, the problem can't be resolved at the infrastructure level; therefore, NServiceBus users must analyze their systems, consider how they are expected to evolve, and define the strategy which will make the most sense in their particular circumstances.

This article presents basic guidelines for choosing a contract evolution strategy, avoiding common mistakes, and ensuring that contracts will be easy to evolve over time.

TODO: link to new document
Note: Ensure that message contracts follow the general [messages design guidelines](/nservicebus/messaging/messages-events-commands.md#designing-messages).

## Adding new message contracts

> When adding new message contracts to a contracts assembly, update the endpoints receiving or subscribing to the new contract type first. If the sender/publisher endpoints are updated first, receivers and subscribers will not process messages for the new contract type.

Author: * Events: usually update publisher first and then subscribers can subscribe at any time. (if there is a requirement to receive all events, subscriber should be updated first)
* Commands: Upgrade the receiver first, otherwise the receiver would have "handler not found" exceptions

Author: this approach can be used for completely new messages but also for "new versions" that are just published twice.

> ## Adding data to existing contracts

Author: "data" is ambiguous? Use properties instead?

TODO: what about version namespaces?

### Add new properties to existing contract types

1. Add the new properties to the existing type.
1. Update senders to the new contract version, setting the additional data on the message
1. Update receivers to handle the new contract version.

This approach requires less modifications to the contracts and even allows to update senders and receivers in arbitrary order as long as the receivers can handle the absence of a value for the new property.
TODO: can serialzers handle the absence?  

It is recommended to use nullable types for the new properties to allow receivers to identify whether it is dealing with an old version of the contract.

### Use inheritance to create a new sub-type

1. Create a new contract type, inheriting from the contract type that should be extended
1. Add the new properties to the new sub-type
1. Update the senders to publish/send the new sub-type
1. Update the receivers to the new sub-type

This approach requires all senders/publishers to be updated first. Due to message inheritance, receivers continue to process the original message version.

> ## Removing contracts or properties

Author: maybe document multi-inheritance as a solution as well?
allows to have contracts with "contradicting" properties but the publisher only has to do a single publish
